### DISCUSSION TITLE
gatsby-plugin-image, getting error: Cannot query field "childImageSharp" on type "ContentfulAssetFile".

### DISCUSSION THREAD

**Discussion by oleksii-lukin** [gatsbyjs/gatsby#30657] [Help] [answered]
It works fine with `file` (from examples if I recall correctly)

```graphql
{
  file(relativePath: {eq: "shared/glasses-image.png"}) {
    childImageSharp {
      gatsbyImage(layout: FLUID, maxWidth: 664) {
        imageData
      }
    }
  }
}
```

but with this query

```graphql
query allContentfulProductsQuery {
  allContentfulProducts {
    edges {
      node {
        mainImage {
          childImageSharp {
            gatsbyImage(layout: FLUID, maxWidth: 664) {
              imageData
            }
          }
        }
      }
    }
  }
}
```

or this part changed to file

```graphql
        mainImage {
          file {
            childImageSharp {
              gatsbyImage(layout: FLUID, maxWidth: 664) {
                imageData
              }
            }
          }
        }
```

it's throwing an error `Cannot query field "childImageSharp" on type "ContentfulAsset".`

Can I use `gatsby-plugin-image` with contentful assets images? Or should it have some particular type selected in content model or inside config of `gatsby-source-contentful` or `gatsby-plugin-sharp` or other plugin I may be missing?

**Comment by rodneylab** [comment, marked as answer]
Hi Alexey,
I think you might need to try querying a `localFile` object.  For the Gatsby 100 Days of Code challenge with Contentful I had something like this:

```
export const query = graphql`
  query LocationTemplateQuery($id: String) {
    site {
      ...SEOFragment
    }
    contentfulLocation(id: { eq: $id }) {
      bannerImage {
        localFile {
          childImageSharp {
            gatsbyImageData(
              width: 1984
              layout: CONSTRAINED
              sizes: "(max-width: 480px) 480px, (max-width: 768px) 768px, 992px"
            )
          }
        }
      }
    }
```
(see [this repo](https://github.com/rodneylab/rodneylab-100-days-of-gatsby/blob/main/src/pages/location/%7BContentfulLocation.slug%7D.jsx) for more context)

Can you adapt this query for you case, to something like:
```
query allContentfulProductsQuery {
  allContentfulProducts {
    edges {
      node {
        mainImage {
          localFile {
            childImageSharp {
             gatsbyImage(layout: FLUID, maxWidth: 664) {
                imageData
              }
            }
          }
        }
      }
    }
  }
}
```
Tell me if I am talking nonsense and I can take another look!

Rodney

**Reply by oleksii-lukin** [reply]
And where that `localFile` coming from? Is it result of `gatsby-source-contentful` operation or another plugin? I don't have such in my GraphiQL. `file` I used above is of type `ContentfulAssetFile`, what is the type of that `localFile`? May be this part which I missing `downloadLocal: true,`... I probably will find it in your repo, going to check it in details later, thank you!

**Reply by oleksii-lukin** [reply]
Oh yes, it probably is

> downloadLocal [boolean][optional] [default: false]
> Downloads and caches ContentfulAsset’s to the local filesystem. Allows you to query a ContentfulAsset’s localFile field, which is not linked to Contentful’s CDN. See Download assets for static distribution for more information on when and how to use this feature.
>

**Reply by rodneylab** [reply]
Hi Alexey,
Yes localFile was in my GraphiQL I didn't have to do anything special to get it to appear.  Search around a little in the tree, it might be hiding somewhere deeper!
I used the gatsby-source-contentul plugin version 5.2.0.
Let me know how you get on

**Reply by k31tha** [reply]
If you use the explorer url ie http://localhost:8000/___graphql what do you see when you navigate?
Im using for example
`query MyQuery {
  allContentfulLocation {
    edges {
      node {
        heroImage {
          gatsbyImageData
        }
      }
    }
  }
}`

**Reply by oleksii-lukin** [reply]
@k31tha I see no `gatsbyImageData` on assets.

I tried example from https://github.com/gatsbyjs/gatsby/tree/master/examples/using-contentful
and it have it on assets:
![image](https://user-images.githubusercontent.com/1281970/113569612-77048380-961b-11eb-82e9-859060917e98.png)

while for my site (using same plugins `gatsby-plugin-image`, `gatsby-source-contentful`, `gatsby-transformer-remark`) I see no `gatsbyImageData` on `ContentfulAsset` nodes. There is `gatsbyImageData` on file nodes, but it's not present for `ContentfulAsset` nodes

![image](https://user-images.githubusercontent.com/1281970/113569954-09a52280-961c-11eb-9547-ae1bab283fcd.png)

**Reply by oleksii-lukin** [reply]
I think I found what the problem was. My `gatsby` were still at `2.xx`, after I updated gatsby and all plugins to `next` versions, it's now appeared.
